pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -659,7 +663,41 @@ func podAffinity(i *infinispanv1.Infinispan, matchLabels map[string]string) *cor
 	return i.Spec.Affinity
 }
 
-// deploymentForInfinispan returns an infinispan Deployment object
+func (r *ReconcileInfinispan) updatePodsLabels(m *infinispanv1.Infinispan, podList *corev1.PodList) error {
+	if len(podList.Items) == 0 {
+		return nil
+	}
+
+	labelsForPod := PodLabels(m.Name)
+	m.AddOperatorLabelsForPods(labelsForPod)
+	m.AddLabelsForPods(labelsForPod)
+
+	for _, pod := range podList.Items {
+		podLabels := make(map[string]string)
+		for index, value := range pod.Labels {
+			if _, ok := labelsForPod[index]; ok || consts.StatefulSetPodLabels[index] {
+				podLabels[index] = value
+			}
+		}
+		for index, value := range labelsForPod {
+			podLabels[index] = value
+		}
+
+		_, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, &pod, func() error {
+			if pod.CreationTimestamp.IsZero() {
+				return errors.NewNotFound(corev1.Resource(""), pod.Name)
+			}
+			pod.Labels = podLabels
```

Comment:
Does this wipe out all the labels from the pod and then restore only our stuff? If so, it's too invasive imo

Reply:
No, the system labels controlled by `consts.StatefulSetPodLabels` and never been removed

Reply:
I'm referring to labels possibly added by the user for his own business, maybe I'm misreading the code but looks like they are not preserved.
also the `cordinator` label for cross site is removed if present.
I would apply a patch policy instead of rebuilding the labels map from scratch

Reply:
Good catch related coordinator label for Pods, thanks @rigazilla . Will merge it logic with current update. Not sure that customers will need own labels on Pods controlled by statefulset, they will be removed every rolling upgrade. If customers will need own labels, they will need to set them via Infinispan CR.

Reply:
> I would apply a patch policy instead of rebuilding the labels map from scratch

Also is unable to implement two-way merge without creating labels map from scratch, but it seems to be similar logic

Reply:
@rigazilla fixed issue with coordinator Pod label

Reply:
I still prefer my idea of not actively delete unknown labels, but that's just a detail, overall LGTM.

Reply:
@rigazilla yes, sure, but I can't imagine how to remove user defined label which user was removed.

Reply:
oh sorry @dmvolod I didn't though to that problem. I got your point now, your are right.

Actually I can something like: if the label name is listed in the annotations but not in the infinispan CR labels, then delete the label on the pod/svc.
this complicates things a bit and I'm not sure if it has side effects.

Reply:
Yes, that's possible solution, but can't handle edit annotations.

Reply:
yeah fine. I think we can go with this solution, let's keep overcomplicated solutions if users will complain about disappearing labels.